examples/todomvc/src/todo.rs

```diff
@@ -0,0 +1,229 @@
+//! Type definition and `dodrio::Render` implementation for a single TODO item.
+
+use crate::keys;
+use dodrio::{bumpalo::Bump, on, Attribute, Node, Render, RootRender, VdomWeak};
+use serde::{Deserialize, Serialize};
+use std::marker::PhantomData;
+use wasm_bindgen::{prelude::*, JsCast};
+
+/// A single TODO item.
+#[derive(Serialize, Deserialize)]
+pub struct Todo<C> {
+    id: usize,
+    title: String,
+    completed: bool,
+
+    #[serde(skip)]
+    edits: Option<String>,
+
+    #[serde(skip)]
+    _controller: PhantomData<C>,
+}
+
+/// Actions on a single TODO item that can be triggered from the UI.
+pub trait TodoActions {
+    /// Toggle the completion state of the TODO item with the given id.
+    fn toggle_completed(root: &mut dyn RootRender, vdom: VdomWeak, id: usize);
+
+    /// Delete the TODO item with the given id.
+    fn delete(root: &mut dyn RootRender, vdom: VdomWeak, id: usize);
+
+    /// Begin editing the TODO item with the given id.
+    fn begin_editing(root: &mut dyn RootRender, vdom: VdomWeak, id: usize);
+
+    /// Update the edits for the TODO with the given id.
+    fn update_edits(root: &mut dyn RootRender, vdom: VdomWeak, id: usize, edits: String);
+
+    /// Finish editing the TODO with the given id.
+    fn finish_edits(root: &mut dyn RootRender, vdom: VdomWeak, id: usize);
+
+    /// Cancel editing the TODO with the given id.
+    fn cancel_edits(root: &mut dyn RootRender, vdom: VdomWeak, id: usize);
+}
+
+impl<C> Todo<C> {
+    /// Construct a new `Todo` with the given identifier and title.
+    pub fn new<S: Into<String>>(id: usize, title: S) -> Self {
+        let title = title.into();
+        let completed = false;
+        let edits = None;
+        Todo {
+            id,
+            title,
+            completed,
+            edits,
+            _controller: PhantomData,
+        }
+    }
+
+    /// Set this TODO item's id.
+    pub fn set_id(&mut self, id: usize) {
+        self.id = id;
+    }
+
+    /// Is this `Todo` complete?
+    pub fn is_complete(&self) -> bool {
+        self.completed
+    }
+
+    /// Mark the `Todo` as complete or not.
+    pub fn set_complete(&mut self, to: bool) {
+        self.completed = to;
+    }
+
+    /// Get this TODO's title.
+    pub fn title(&self) -> &str {
```

Comment:
I smell a derive macro for stuff like this

Reply:
or just make it all `pub`... a real application would probably not use `pub` so I didn't here either. This is kind of an in-between scenario.